### ISSUE TITLE
We need to define the JPMS strategy for EE10 and beyond

### ISSUE THREAD

**Issue by starksm64** [eclipse-ee4j/jakartaee-platform#329]
**Is your feature request related to a problem? Please describe.**
We have a problem with inconsistent JPMS usage for the platform APIs jars.

**Describe the solution you'd like**
We need to clearly define the requirements for supporting JPMS.

**Describe alternatives you've considered**
The current approach of allowing specification projects to define the level of JPMS support has produced an inconsistent experience for users/servers attempting to use JPMS.

**Comment by starksm64** [comment]
Previous issues:
#174

**Comment by starksm64** [comment]
Ideas on various levels of details that are needed based on internal discussions:

* Simply establishing a module naming convention for Jakarta EE APIs
* Requiring automodules for Jakarta EE APIs which expose those names
* Requiring (or allowing) actual module descriptors for Jakarta EE API artifacts
* Requiring specific content within module descriptors for Jakarta EE API artifacts (for example, version information, annotations, dependency information, etc.)
* Requiring (or allowing) Jakarta EE API artifacts to be deployed on the module path or in another JPMS-compatible environment
* Requiring (or allowing) programs which consume Jakarta EE API artifacts to be deployed on the module path or in another JPMS-compatible environment
* Establishing behaviors or conventions around usage of module mechanisms in a Jakarta EE environment (for example, special behavior specifications around service loading)

**Comment by keilw** [comment]
It may be a special niche but based on Bugs at least until Java 13 or 14 were not addressed (so if the minimum JDK version was 11-13 there is a great risk of that) JPMS and Multi-Release-JARs are incompatible and may cause serious problems if used together without safety-measures. 

- We demonstrated it in those [demos for JSR 385](https://github.com/unitsofmeasurement/uom-demos/tree/master/console/java13)
-  This issue for the [JSR 385 RI](https://github.com/unitsofmeasurement/indriya/issues/301) could only be solved by adding **both** a `module-info` (especially in a Multi-Release-JAR structure) and an `Automatic-Module-Name` as a fallback, since Java 11 (and potentially still 12, it looks resolved after 13) can ignore the Multi-Release-JAR and simply call the initial structure if JPMS and Multi-Release-JAR get combined

**Comment by tomas-langer** [comment]
Is there a plan to upgrade minimal Java version requirement? Considering Java 8 is end of life, if we base Jakarta 10 on Java 11, we do not need to have multi-release jars, and can fully utilize JPMS using `module-info.java`.

**Comment by keilw** [comment]
That's a good point, I am also not entirely sure, if OpenJDK some day pruned or eliminated the `ServiceLoader` mechanism via `META-INF\services` only to use definitions in `module-info.java`?

Should any project plan to use "fancy new stuff" like records, etc. (maybe not all of it likely for an API but who knows what they plan to use) then even with a minimum Java SE version like 10, 11 or 12 there could still be a temptation and need to use these via Multi-Release-JAR although I must admit, for specs there may not be such a benefit, given that for the time being I don't think a spec may offer element X but Y only if you are on Java 16 or above? ;-)

**Comment by anthonyvdotbe** [comment]
tl;dr Java SE 11 for module descriptors (and Java SE 8 or 11 for source code, but that question is irrelevant w.r.t. modules)

@keilw I don't see any evidence that proves there are unresolved JDK issues w.r.t. the module system and/or MR JARs:
* the issue with JSR 385 is that Indriya's MR JAR does not comply with the specification (I've filed https://github.com/unitsofmeasurement/indriya/issues/346 for this)
* you say "OpenJDK Bugs like JDK-8207162 and similar ones", but there's not a single link to an unresolved issue to back up your claims ([JDK-8207162](https://bugs.openjdk.java.net/browse/JDK-8207162) was already fixed in JDK 12 and backported to JDK 11.0.2. Moreover, it's `jdeps`-specific and thus irrelevant anyway)
* even if there would be existing or yet-undiscovered issues: Java SE 11 is the only LTS release to date with module support, and thus also the only one for which backports would be done

@tomas-langer MR JARs are not a prerequisite for Java SE 8-compatible modular JARs: the `module-info.class` can simply be placed in the JAR's root directory

> I don't think a spec may offer element X but Y only if you are on Java 16 or above?

This question is irrelevant, because the JAR spec itself doesn't currently allow it:

> The public API exported by the classes in a multi-release JAR file must be exactly the same across versions [...] A future release of this specification may relax the exact same API constraint to support careful evolution.

**Comment by keilw** [comment]
@anthonyvdotbe Thanks but unitsofmeasurement/indriya#346 is meaningless and I already closed it as invalid.
Frankly I don't care if the JDK team did not do their homework or papertrail. What I and users care is that it works and it doesn't between Java 9 and 12, there is neither a JDK 11.0.2 this works with (we tested it, it fails with every JDK 11 up to jdk-11.0.8) nor any released Java 12, it was only fixed from 13 and they simply forgot or were too busy with other things to backport it. Or it was only made available to paying customers but not the common OpenJDK builds available under AdoptOpenJDK or similar.

**Comment by lukasj** [comment]
> Requiring automodules for Jakarta EE APIs which expose those names

This should be no-go for spec projects. There are more problems with them but the one I mean here is that `jlink` does not support them. Even though that can be seen as not-an-issue from the EE/appserver perspective, it can be problem in various standalone implementations and there were users who were quite loud about this in few projects (ie activation can be an example) and Jakarta EE definitely wants to allow and also support usage of various specifications out of the EE container.


> Establishing behaviors or conventions around usage of module mechanisms in a Jakarta EE environment (for example, special behavior specifications around service loading)

open module propagation to allow reflection access is another thing. What I as an end user want is to `open pkg.entities to jakarta.persistence` and let the persistence API figure out to which implementation this `open` should be propagated to. I do not want to rely on particular persistence implementation be it EclipseLink or Hibernate or whatever else. This can be generalized as sth like _To allow reflection access from the CI to user defined classes, user code must be open to `specJar` which propagates this openess to an implementation found by lookup mechanism defined in `spec`_ (where _user defined classes_ can be ie entities, jaxb/jsonb annotated classes etc)


> Is there a plan to upgrade minimal Java version requirement? Considering Java 8 is end of life,..

Do not take me wrong from the text bellow - I vote for SE 11 as the base should my vote be the binding one ;-)
While Java SE 8 is EOL, support for it is going to end by December 2030, so in almost 10 years from now. I'd be personally very happy to jump directly over to SE 17 right now all over the place but I think, EE 10 should still run on 8 and min Java SE version should move to 11 in Jakarta EE 11. (Should there be any 10.X releases, all of them should keep the same min Java SE level as initial 10.0 version - but I think this is clear).


> if OpenJDK some day pruned or eliminated the ServiceLoader mechanism via META-INF\services only to use definitions in module-info.java? From what I know, if you define them as provides in the latter you should not need to define the services folder as well (unless you need a fallback for JDKs before 9)

https://docs.oracle.com/en/java/javase/11/docs/specs/jar/jar.html#modular-jar-files gives clear answer to that:
>  If the modular JAR file is deployed on the class path then it behaves as if a non-modular JAR file.

Which, in the other words means that `module-info` is ignored if the jar is on the classpath. So as long as class-path mode is to be supported, `META-INF/services` must be present.


> Should any project plan to use "fancy new stuff" like records, etc. 

At the spec project level, usage of  "fancy new stuff" should be discouraged (or not recommended) - but probably not strictly prohibited, unless it comes from an LTS Java SE version (ie support for SE 15 would not be recommended, for SE 17 it would be OK). My thinking here is that implementations/vendors should not be forced to provide support for possibly changing features. If they want to support them in their implementations, they can do so even under current rules.


>  the module-info.class can simply be placed in the JAR's root directory

or under the versioned directory. It is possible to have slightly different descriptors for different SE versions. This is defined at https://docs.oracle.com/en/java/javase/11/docs/specs/jar/jar.html#modular-multi-release-jar-files - this also proves that the statement *JPMS and Multi-Release-JAR mechanism are incompatible* is *false* as long as defined rules are followed.



I'm not sure it is feasible to have this completely covered, defined and implemented in EE 10. But it would be definitely good to start with _spec and CI can support JPMS_ by following TBD rules/recommendations and switching it over to _spec must and CI can support JPMS_ in EE 11 together with further adjustments and bumping up min SE version to 11 there. It would be nice to say _must_ also for the CIs in EE 11 but I'm afraid there will be implementations not willing to jump on the JPMS train.

**Comment by nipafx** [comment]
Hi there, sorry to barge in but I heard people are discussing modules and I'm always up for that. :) Caveat: I don't know the Jakarta EE space well, so I may be missing context - let me know if you think I do.

A few random replies:

> We need to clearly define the requirements for supporting JPMS

👍🏾  Maybe [this helps](https://nipafx.dev/java-modules-jpms-maturity-model/). (Note that it has no official character and didn't catch on, but I still think it's good. 😊 )

> Requiring automodules for Jakarta EE APIs which expose those names

Technically, the automatic module name is only that: a name for the JAR if it's used on the module path. But socially, it also signals to users that the JAR works fine on the module path. I recommend to only define a name once that has been established.

> It may be a special niche but based on Bugs at least until Java 13 not fully addressed (so if the minimum JDK version was 11-13 there is a great risk of that) JPMS and Multi-Release-JAR mechanism are incompatible and may cause serious problems if used together without safety-measures.

As far as I can tell, this analysis is based on a malformed JAR (see my comment in unitsofmeasurement/indriya#346) and unrelated jdeps and Maven issues. I don't think it holds up.

> I am also not entirely sure, if OpenJDK some day pruned or eliminated the `ServiceLoader` mechanism via `META-INF\services` only to use definitions in `module-info.java`?

I have no particular insight into plans for the service loader, but I'd be extremely surprised by such a change as it would severely neuter the class path. So unless the class path itself is going away (would be equally surprising), I wouldn't worry about this. And @lukasj got it right re service declarations: If a JAR is supposed to work on class path and module path, it needs to define services in `META-INF\services` and `module-info.java`?

> What I and users care is that it works and it doesn't between Java 9 and 12, there is neither a JDK 11.0.2 this works with (we tested it, it fails with every JDK 11 up to jdk-11.0.8) nor any released Java 12, it was only fixed from 13 and they simply forgot or were too busy with other things to backport it. Or it was only made available to paying customers but not the common OpenJDK builds available under AdoptOpenJDK or similar.

Re fixes in 13 not being backported in 12, that's because 12 is no LTS version. As far as I can tell nobody in the OpenJDK community is backporting fixes to it and neither does anybody offer commercial support for it.

> At the spec project level, usage of "fancy new stuff" should be discouraged (or not recommended) - but probably not strictly prohibited, unless it comes from an LTS Java SE version (ie support for SE 15 would not be recommended, for SE 17 it would be OK). My thinking here is that implementations/vendors should not be forced to provide support for possibly changing features.

I may well have misunderstood the last sentence, so this might be a non-sequitur, but I want to point out that only preview features (enabled with `--enable-preview`) will change between releases. Once they're out of preview (e.g. text blocks in 15), they're set in stone regardless of whether that version is LTS or not. More to the point LTS is a commercial concept that the JLS doesn't give [two Guy Fawkes](https://www.youtube.com/watch?v=Yow_BJeb8TI&t=70s) about. 😁

**Comment by keilw** [comment]
@nipafx Thanks for your input here. I just literelly at this moment saw the mail, that DWX' 21 also has to happen remotely, so while I anticipate your JDK talk, we (including @thodorisbais, @ivargrimstad, you and myself) won't meet in person then, because after the recent Corona "Emergency Break" in Germany they certainly did not see a chance to hold it in person, especially not in the home town of Markus Söder ;-/

You made most of the good points here, especially about `--enable-preview` which the initial JSR 376 spec (that remained unchanged ever since a vast number of us here who were also in the JCP EC back then voted against it in 2017 and some, but only minimal changes were applied but enough to get us approve it in the end) did not even know back then, but especially problems and bugs related to jdepts or jlink like [JDK-8207162](https://bugs.openjdk.java.net/browse/JDK-8207162) are going to be a problem as long as the mimimal Java version stays at 9-11 or even 12.

@anthonyvdotbe Sorry to dissapoint you, but I was also quite surprised to even learn of your ticket #174 which only @kwsutter had ever exchanged a few thoughts with until @starksm64 linked it here and some had a look at it. I was quite actively involved in shaping a minumum level of modularity and JPMS in Jakarta EE 9 already, but there were some compromises (as everywhere also in other specs especially in the JDK itself ;-) and a few specs or APIs that did not feel ready at the time especially CDI while others had done that already.
It seems you made at least one commit to [microprofile](https://github.com/eclipse/microprofile/graphs/contributors), have you ever tried approaching it about JPMS because while 2/3 or more of Jakarta EE specs and APIs already support at least Automatic-Module-Name since Jakarta EE 9 or earlier, there is zero evidence the official MP APIs ever got a single JPMS definition of their own, the only one seems to be added later by projects like Helidon (@lukasj probably knows a lot more about that) and unfortunately not even some key stakeholders to Eclipse MicroProfile are always completely aware of using JPMS and problems that may arise if you do it wrong, especially **Split Packages**? Did you ever raise those kinds of tickets with MicroProfile like you did here and we just don't know about or did you not suggest it there @anthonyvdotbe, I think while I cannot promise it'll succeed with every single contributor you should try to raise awareness there if you care about this topic.

**Comment by pron** [comment]
>  support for it is going to end by December 2030, so in almost 10 years from now

Support is going to continue *at least* until December 2030. Oracle might decide to extend it further as long as there are enough paying customers. 

The question is, what does this have to do with support for Java 8 in *new* versions of libraries/specifications? Java 8 is essentially EOL and has been very much discouraged for *new* projects for a while, now. It is now in pure legacy maintenance mode, and new specifications targeting it might confuse people further about this.

**Comment by starksm64** [comment]
> Is there a plan to upgrade minimal Java version requirement? Considering Java 8 is end of life, if we base Jakarta 10 on Java 11, we do not need to have multi-release jars, and can fully utilize JPMS using `module-info.java`.

The expectation is that at least Java 11 will be required. This is being tracked in #331 where the question of Java 17 support is also raised.

**Comment by keilw** [comment]
@pron For completely new services I agree and I also helped a customer introduce a green-field system based on Java 11+ around 2 years ago, but right now I have another case where "Legacy" systems like Tibco, CORBA or even Host systems are involved in a very big company and none of those are being replaced before at least the mid 2020s. Thus they add new services using Java 11 and Jakarta EE or Spring Boot, etc. but they still have those old systems and for SOAP or other protocols many don't get touched until eventually they might be replaced by something like a new REST based API and only then when all the services are upgraded, the last Java 8 client or server is going to disappear. 
And keep in mind, Java 8 has a longer Extended Service guarantee than even the upcoming Java 17, see [Oracle Java SE Support Roadmap](https://www.oracle.com/java/technologies/java-se-support-roadmap.html), so some customers including the ones I deal with at the moment are more likely to switch from 8 to 23 in Fall 2024 when that seems like the next LTS following a 3-year cycle.

**Comment by pron** [comment]
@keilw Java 8 will be supported for as long as Oracle deems it worthwhile; could be 2050; could be 2150. The point is that while maintenance releases of existing specifications, like Jakarta EE 9, might want to support Java 8 until, say, 2045, this shouldn't affect the minimum required version for *new* specifications, like Jakarta EE 10. I mean, you can choose to target Java 8 if you like, but my point is that Oracle's legacy support for 8 shouldn't impact that decision one way or another. Java 8 should no longer be used for any new projects, and new specification versions can safely (and probably should) choose not to support it.

**Comment by keilw** [comment]
@pron Sure but then it becomes expensive. I worked at BEA just before Oracle purchased it and many of those support plans for the Middleware still exist or were combined with what Oracle had. And we had Weblogic 4 customers (2007 WLS 10 was released, so 4 existed since 1999) who needed support by the Professional Suppot team I helped and paid somewhere between a single and double-digit Million $ amount per year for that, so yes, if you are happy to pay you could use it until the end of times.

**Comment by pron** [comment]
My point is just that Oracle's (possibly eternal) support period should not play a factor in the decision of which version to support. Java 8 is EOL for new work and there's no need to support it in new spec versions.

**Comment by starksm64** [comment]
> My point is just that Oracle's (possibly eternal) support period should not play a factor in the decision of which version to support. Java 8 is EOL for new work and there's no need to support it in new spec versions.

Agreed. The discussion on the minimal Java SE version should occur on #331. Currently there is no expectation that the Java SE version will be less than 11.

**Comment by lukasj** [comment]
> My point is just that Oracle's (possibly eternal) support period should not play a factor in the decision of which version to support.

Point taken. My concern is "What's next, after EE 10 gets out, and when?" What I would expect from the Platform project is at least rough, say mid-term, schedule for at least current + 1 releases in terms of supported Java SE versions. I could have probably missed or haven't heard ie sth like: _Within 6 months of new Java SE LTS release X there is new Jakarta EE Platform release with Java SE X set as a minimum supported Java SE version. This version of Jakarta EE allows changes breaking backwards compatibility rules._ and _Every 6-12 months there may be an update release for the latest Jakarta EE version available. This version does not allow breaking changes and requires strong backward compatibility with the EE version the update is based on._  That would send clear message not only to spec projects, which can then plan features for the platform as a whole and/or their updates accordingly, but also to the community/customers wrt when there will be an update/new major version and what its basic requirements are going to be.

**Comment by keilw** [comment]
This has been a little silent, maybe it'll gain more speed after 9.1 is out, but one important finding after a little "goose chase" we finally got an answer to what happened in https://github.com/unitsofmeasurement/indriya/issues/301 although the answers came from another completely unrelated issue ticket.

**Comment by anthonyvdotbe** [comment]
tl;dr [JDK-8235229](https://bugs.openjdk.java.net/browse/JDK-8235229) is irrelevant, both to this issue and #331

JDK-8235229 was a compilation-only issue: when using the `--release` parameter, `javac` didn't treat multi-release JARs as such, and only considered their top-level directories.

Since the JAR spec specifies that:
> The public API exported by the classes in a multi-release JAR file must be exactly the same across versions

it had a negligible impact (which is why it wasn't reported until after the release of Java SE 13 and is unlikely to be backported). The only reasonable scenario (i.e. a scenario that only involves spec-compliant multi-release JARs) I can readily think of is the one in the report: a modular multi-release JAR that does not contain a `module-info.class` in the top-level directory, but does contain one in the `9`-versioned directory. In this case, the JAR is incorrectly treated as an automatic module, which might result in erroneous - but harmless - warnings being issued. (To eliminate those warnings, one can either compile with JDK 16+ and specify the appropriate `--release` parameter, or compile with the appropriate JDK and not specify the `--release` parameter.)

**Comment by keilw** [comment]
Well compilation is bad enough, so there needs to be a warning in certain documents (probably user guide at the very least) so even if the combination of modules and multi-release-JARs may still be rare people are aware and know what to do. 
With Java SE 16 final they can find a way around it as long as the application has no problems with Java 16. And #331 must go up to Java 16, otherwise they could not use it yet if the compatible Java version range was too narrow.

**Comment by anthonyvdotbe** [comment]
tl;dr I agree that JDK-8235229 is relevant, in the sense that Jakarta EE should not produce any such JARs (by requiring a `module-info.class` file to be present in the top-level directory)

About the lack of `Automatic-Module-Name`: you're right, it will cause compilation errors if the filename-derived module name doesn't match the one in the module descriptor.
About `META-INF/services` being absent or inconsistent: I'd argue that's a malformed JAR. If you have multiple module descriptors in an MR JAR, their `provides` clauses must be the same. So since `META-INF/services` is really just "pre-modular `provides` clauses", it should be consistent with any module descriptors' `provides` clauses as well.
About all packages being exposed: you're right, this might cause an inconsistency, where compilation succeeds but an exception is thrown at run-time. However, it's likely that IDEs handle this kind of MR JARs correctly already, thereby preventing this from happening in practice.

So again, the issue can only occur in a very specific case: a modular multi-release JAR that does not contain a `module-info.class` in the top-level directory, but does contain one in the `9`-versioned directory. And for it to cause trouble in practice, even more conditions must be satisfied. Either way, neither `junit-platform-commons` (which has one in the top-level directory) nor `log4j-core` (which doesn't have one in the `9`-versioned directory) are problematic.

Also, there are several ways to deal with it. As mentioned, one can either compile with JDK 16+ and specify the appropriate `--release` parameter (this does not require to run Maven itself with such JDK), or compile with the appropriate JDK and not specify the `--release` parameter (I'd expect this to be a trivial change in most cases).
Another option is even simpler: do nothing. The worst-case scenario here is that some code relies on non-exported packages *and* that none of the tests actually test this code. But at this point there's so many conditions that must be satisfied, that I doubt this will ever occur in practice, especially in the context of Jakarta EE.

**Comment by keilw** [comment]
No `META-INF/services` is not even required in a modular world, take Helidon and the colleagues here can probably tell even more which `provides io.helidon.config.spi.ConfigParser with PropertiesConfigParser;`
and has no trace of a `META-INF/services` doing the same.
Now Helidon does not use a multi-release JAR but Jakarta EE cannot control or enforce what any given third party library does out there and issue reports like the mentioned (only as brief example) show these happen in the "Cloud Native Java" World including Micronaut, Spring, MicroProfile or all the other "Microframeworks" most of which also build on top of Java or Jakarta EE.

**Comment by lukasj** [comment]
> No `META-INF/services` is not even required in a modular world

Jakarta MUST keep compatibility with the classpath mode. In that world, `META-INF/services` are mandatory. Wrt Helidon, it started to create entries in `META-INF/services` during the build based on the content in `module-info` (there is a maven plugin which does that)

**Comment by keilw** [comment]
@lukasj Thanks for the input. I guess where backward-compatibility of the API even down to Java 8 is required for the time being (likely up to another decade) Jakarta EE has higher backward compatibility requirements than some of the "Microframeworks" may have.

Of course the `module-info` also does other things especially hiding internal packages with local helper classes used only by the system and not exposed anywhere.
It may be a very rare case as well, but if in a multi-release JAR there's a need for more platform-specific internal packages or classes, access to those can be regulated in the module-info in various ways also ensuring that `ClassA` was only visible to `ClassB` and nothing else. Falling back to a root-level module-info or even worse `Automatic-Module-Name` those contstraints are lost and classes or packages not intended to be used on the outside could be visible.

**Comment by lukasj** [comment]
My comment is not necessarily related to backwards compatibility as such. As long as Java SE itself supports class-path, there is no way for jakarta APIs to deny people putting APIs there effectively bypassing all constraints defined in `module-info` regardless of where exactly the descriptor is located.

**Comment by keilw** [comment]
Of course the APIs should and while it is up to what individual vendors and their compatible implementations do (if they force them to upgrade to Java 17 it's their call) the former RIs like Glassfish may also go a little softer here and e.g. try to avoid multi-release JARs or offer enough backup mechanisms if they were not used yet.

**Comment by lukasj** [comment]
> the former RIs like Glassfish may also go a little softer here and e.g. try to avoid multi-release JARs 

6.1 contains few MR Jars already (with module-info in place - in jar's root), 6.0 as well, and I haven't heard about any issues on 11 caused by this yet

**Comment by keilw** [comment]
I think if you keep module-info in the root that's normally not affected, but in more complex systems there may well be multiple `module-info` files in addition to that and those may e.g. fall-back to the default if you have it. And it is spec-compliant to "also" put the module-info into a Java 9 folder.

>A module descriptor is generally treated no differently to any other class or resource file. A module descriptor may be present under a versioned area but not present under the top-level directory. This ensures, for example, only Java 8 versioned classes can be present under the top-level directory while Java 9 versioned classes (including, or perhaps only, the module descriptor) can be present under the 9 versioned directory.

Even suggests that might be the preferred way.

**Comment by anthonyvdotbe** [comment]
> the issue can only occur in a very specific case: a modular multi-release JAR that does not contain a module-info.class in the top-level directory, but does contain one in the 9-versioned directory.

I invite everyone to find an actual library like this. (I'm not saying there aren't any, just that I predict it won't be easy to find one, let alone one that is somewhat popular.) Also note that there's likely tools/libraries out there that support modules, but don't support MR JARs yet. So it's equally likely that module descriptors are being placed in the top-level directory already in order to accommodate those.
And as explained before, the implications in practice are negligible (which is why it went unnoticed until after Java SE 13, was classified as a P4 bug, and there hasn't been any fuss about it until [3 days ago](https://github.com/unitsofmeasurement/indriya/issues/346#issuecomment-836661541)).

**Comment by keilw** [comment]
Well it's not up to the Jakarta EE WG to "find libraries" and the spec allows both, plus there can be multiple versioned directories (the example only mentions one, but the [bug 301](https://github.com/unitsofmeasurement/indriya/issues/301) filed gainst Indriya earlier also was raised for a **true multi-release JAR**) with more than just one, that I agree is probably even more rare but not impossible and there are so many JDK versions between 9 and 15 all of which could face this.
The correct cause of 301 was just identified around 3 days ago, but the actual issue as small as the JDK team may consider it was known much longer and 301 was also filed almost a year ago.

**Comment by lukasj** [comment]
this is going nowhere. The responsibility of the WG/platform group should be to say that it wants/does not want to support JPMS and if it wants to support it, then define requirements for the specs to be satisfied in order to be included in the platform X.Y.Z and approximate time-line for delivering full featured JPMS support. I don't think it is feasible to have everything ready within one release. I think it would be sufficient to include this as a (strong?) recommentation in the platform architecture document @ivargrimstad created.

Platform may also require support for JPMS  by spec's CIs at some point in the future.

I can imagine that for EE 10 requirements on included specs can be:
* spec module name is `jakarta.<specName>`
* spec provides JPMS module definition (reliance on automatic module name is not allowed)
* spec updates provider lookup algorithm to include search on `module-path` (if applicable)

While this certainly does not capture everything, it should be enough to allow people using specs on JPMS and give community option to come up with new CIs fully utilizing the module system.

If some spec decides (and is allowed to) support multiple newer SE versions through MR jar beyond the Platform defined minimum version, it should be responsibility of that spec/corresponding CIs/platform CIs to resolve any encountered issues - either by not including newer SE version specific support, compiling with latest JDK (fine as long as running on min required platform defined SE level is supported) or working with JDK/other dependencies teams on fixing issues there. Platform project itself as such should not care about these "implementation details" to certain extend.

At the end of the day, what I as an end user want in order to develop my app for the Platform X.Y.Z at some point in the future is to define my app as:

```java
module org.eclipse.platformdemo {
    requires jakartaee.platform;
    requires jakarta.xml.bind;
    requires my.lib;
    exports my.jaxb.classes;
    opens my.jaxb.classes to jakarta.xml.bind;
    ....
}
```

where `jakartaee.platform` comes from `jakarta.platform:jakarta.jakartaee-api:X.Y.Z` artifact, is set to `provided` scope and is defined ie as:

```java
module jakartaee.platform {
    requires transitive jakartaee.web;
    requires transitive jakarta.persistence;
    ....
    // optional parts
    requires static jakarta.xml.bind;
    ....
}
```

This artifact/module contains spec APIs ONLY; no implementation classes should be allowed to guarantee vendor-neutrality (as it is with current version of the `jakarta.jakartaee-api` artifact); should this artifact be put on the classpath, the behaviour is exactly the same as it is today.

Now I want to build my app and "run" it somewhere. Let's say there is vendor V, who opts-in to support JPMS and provides CI for Plarform X.Y.Z. This vendor defines its own implementation of the `jakartaee.platform` module through let's say `com.vendor:jakarta.platform:X.Y.Z` artifact like ie:

```java
module jakartaee.platform {
    requires transitive jakartaee.web;
    requires transitive jakarta.persistence;
    ....
    // we contain optional parts
    requires transitive jakarta.xml.bind;
    ....
    // implementations we use
    requires transitive com.sun.xml.bind;
    requires transitive org.eclipse.persistence.moxy;
    ....
}
```

When my app is deployed into such environment, all modules are properly resolved and my application just runs. Vendor V may also allow deployment which completely ignores modules through some option. There also can be other vendor W, who did not opt-in to support JPMS - such vendor only needs to make sure that all JPMS descriptors are ignored and when I take my app and deploy it into W's environment, the app runs exactly the same way as today.


...just my 2¢...

**Comment by keilw** [comment]
That's precicely what I suggested 2 days ago: https://github.com/eclipse-ee4j/jakartaee-platform/issues/329#issuecomment-838659971 and I don't think we need countless more arguments for and certainly none against it here.
It was on the Agenda for the platform call but this time we did not get there, hopefully in one of the next weeks.

**Comment by lukasj** [comment]
> P.s.: I suppose with "reliance on automatic module name is not allowed" you mean the "fully-automatic" one if nothing is declared in the MANIFEST either?

I originally wanted to say: NO `AUTOMATIC-MODULE-NAME` in the manifest. The difference for most specs which do not have anything yet is whether they add one line to manifest or create new file having ~10 lines if I don't count license header - in terms of time spent on this, how big difference it is? 10 minutes? OTOH when we come down to implementation, it may depend on minimum supported Java SE version by the platform itself and possible issues we may encounter in JDK or tools we use. Therefore I did not want to be that strong on that requirement.

**Comment by starksm64** [comment]
For EE10 we decided on a minimal adoption strategy for API jars with no binding behaviors. This is now an EE11 issue if there are real requirements around supporting JPMS in deployments.
